### DISCUSSION TITLE
Hosting sql files directly inside the database

### DISCUSSION THREAD

**Discussion by pchemguy** [sqlpage/SQLPage#438] [Q&A] [answered]
What do I need to do to host files inside the database apart from placing them in the table as instructed? I use the official binaries for Windows. Can I use it for this purpose? Can I host some files on disk and some in the database? If yes, what is the search logic?

It makes sense to host small files, which are not actively edited in the database, such as images/icons. Larger files or files being actively edited might be better left on disk directly.

**Comment by lovasoa** [comment, marked as answer]
Hello ! 

 - You can host files in the database just by creating and populating the following table
```sql
CREATE TABLE sqlpage_files(
  path VARCHAR(255) NOT NULL PRIMARY KEY,
  contents BLOB,
  last_modified TIMESTAMP DEFAULT CURRENT_TIMESTAMP
);
```
 - Can I host some files on disk and some in the database? Yes
 - If yes, what is the search logic? serve the local file first, if not found, serve from the db

**Reply by amrutadotorg** [reply]
Hi, is there any indicator from the server that the file was taken from the database?

sample log of the file served from the DB:
`[2024-06-21T22:13:21.330Z INFO  actix_web::middleware::logger] 127.0.0.1 "GET /videos/Kundalini%20Rise%20Animation%20with%20music_EN.mp4 HTTP/1.1" 304 0 "https://learnsahajayoga.org/" "Mozilla/5.0 (Linux; Android 10; K) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Mobile Safari/537.36" 0.000202`

**Reply by lovasoa** [reply]
Hi ! You can enable verbose logging by setting the environment variable `RUST_LOG=sqlpage=debug` and you'll see where files are read from in the console. (as well as many other useful debugging information)

**Reply by amrutadotorg** [reply]
Thanks, I do not see it in the logs. Does it mean that the file is being served from the disk?

`[2024-06-22T10:16:17.705Z DEBUG sqlpage::webserver::http] Serving file: "videos/1982-1007_I_bow.mp4"`

```
select path from  sqlpage_files  where path='videos/1982-1007_I_bow.mp4';
            path
----------------------------
 videos/1982-1007_I_bow.mp4
(1 row)
```

```
SELECT 'hero' as component,
'"I bow to all the seekers of truth."' as title,
TRUE as reverse,
'images/video_covers/i_bow.webp' as poster,
'/videos/1982-1007_I_bow.mp4' as video;
```

**Reply by lovasoa** [reply]
The database filesystem setup information should be near the beginning of the logs.

**Reply by amrutadotorg** [reply]
found these 

```
# grep sqlpage_files /var/log/sqlpage.log
    Alternatively, you can create a table called `sqlpage_files` in your database with the following columns: `path`, `contents`, and `last_modified`.' AS contents_md
    Alternatively, you can create a table called `sqlpage_files` in your database with the following columns: `path`, `contents`, and `last_modified`.' AS contents_md optimized to [("contents_md", Static(String("\nAlternatively, you can create a table called `sqlpage_files` in your database with the following columns: `path`, `contents`, and `last_modified`.")))]
```

```
\d sqlpage_files
                              Table "public.sqlpage_files"
    Column     |            Type             | Collation | Nullable |      Default
---------------+-----------------------------+-----------+----------+-------------------
 path          | character varying(255)      |           | not null |
 contents      | bytea                       |           |          |
 last_modified | timestamp without time zone |           |          | CURRENT_TIMESTAMP
Indexes:
    "sqlpage_files_pkey" PRIMARY KEY, btree (path)
```

**Reply by lovasoa** [reply]
If sqlpage cannot access the files table, you should see the following in your logs:

```
Using local filesystem only, could not initialize on-database filesystem. 
```

**Reply by amrutadotorg** [reply]
Thank you. It looks like it is working since I didn't get the message.

I have one more question. If a particular file is not available in the table, will I receive a message about the fallback? I want to ensure that all files are in the table with the correct paths.

**Reply by lovasoa** [reply]
The next version will have more verbose debug logging about file system resolutions

**Reply by amrutadotorg** [reply]
Hi, now the logging looks awesome, thank you!
Is it possible to automatically serve files generated by SQLPages from a table?

for example:
```
[2024-06-25T10:47:10.574Z ERROR sqlpage::webserver::http] Unable to read file "sqlpage.bac98cf4929ec955.js": Unable to read "sqlpage.bac98cf4929ec955.js" from the database: 404 Not Found

[2024-06-25T06:47:21.016Z ERROR sqlpage::webserver::http] Unable to read file "sqlpage.e5b7f4b1dfa3d61c.css": Unable to read "sqlpage.e5b7f4b1dfa3d61c.css" from the database: 404 Not Found
```

**Reply by lovasoa** [reply]
>  Is it possible to automatically serve files generated by SQLPages from a table?

No it's not. SQLPage-generated files are static, they are not meant to be altered, and this allows serving them with a cache policy that allows caching them forever and never requesting them twice. This wouldn't play well with allowing users to overwrite them.
